### PR TITLE
bpo-29741: make some BytesIO methods accept integer types

### PR DESCRIPTION
**Pull request by orenmn**
according to http://bugs.python.org/issue29741:
- change some functions in Modules/_io/stringio.c and in Modules/_io/bytesio.c to use _PyIO_ConvertSsize_t (which supports integer types)
- change _io_BytesIO_truncate_impl so that it would accept integer types
- add tests to test_memoryio
- change stuff in Lib/_pyio.py to behave like the C implementaiton

(I ran the test module again, and on my Windows 10, the same tests failed with
and without my patches. However, on my Ubuntu 16.04 VM, none of the tests
failed.)

https://bugs.python.org/issue29741

### CODE REVIEW

**Comment by orenmn** [comment]
with regard to continuous-integration/appveyor/pr build fail - I see in the logs that test_site failed, but on my Windows 10, test_site passed.
(6 tests did fail on my Windows 10, but they failed both on cpython with and without my patches.)

**Comment by orenmn** [comment]
Does the AppVeyor build fail prevent core devs from reviewing this patch?
And if it does, what can I do about it?

**Comment by orenmn** [comment]
merged changes from PR #606 and PR #650.
that leaves this PR with changes only in Lib/_pyio.py and Lib/test/test_memoryio.py.

File: Lib/_pyio.py
Comment (zooba) [inline Lib/_pyio.py]: Given how `size` gets used after this, there shouldn't be any need to reassign it by calling `__index__`. Similarly for the other changes.

**Review by zooba — COMMENTED** [review]
It seems like with the native module changes the rest of the `_pyio.py` changes may be unnecessary. The enhanced tests are definitely valuable, but it would be nice to see if they pass without adding the extra error messages to _pyio.

**Comment by orenmn** [comment]
I am really busy nowadays (this is my 2nd term), and so I won't be able to work on CPython until August.. if anyone wishes to take it from where I left, please do :)

File: Lib/_pyio.py
Comment (orenmn) [reply Lib/_pyio.py]: when I remove the reassignment, I get `TypeError: '<' not supported between instances of 'int' and 'IntLike'` somewhere later, when the size var is compared to an int.
Did I misunderstand your comment?

Comment (orenmn) [reply Lib/_pyio.py]: Steve?

Comment (zooba) [reply Lib/_pyio.py]: Sorry - GitHub messages go to my work email, which means I don't see them unless I'm sneakily doing Python stuff on work time. Conversation on the bug is better.

There shouldn't be any need for the `from err` - chaining will happen automatically (that syntax is there for explicit non-trivial chaining). It's probably also a good idea to separate the attribute access from the invocation to ensure we wrap up the right exception:

```python
try:
    size_index = size.__index__
except AttributeError:
    raise TypeError(f"{size!r} is not int-like enough")
else:
    size = size_index()   # now AttributeErrors from inside __index__ will be raised cleanly
```

File: Lib/test/test_memoryio.py
Comment (zooba) [inline Lib/test/test_memoryio.py]: No parens after class names

**Review by zooba — CHANGES_REQUESTED** [review]
Just some relatively minor style fixes, but we need to get these right within the stdlib.

File: Lib/test/test_memoryio.py
Comment (orenmn) [reply Lib/test/test_memoryio.py]: my bad.
BTW, do you think this should be added to PEP8, or is it obvious?
(i grepped and found only 7 places in the codebase doing this.)

Comment (zooba) [reply Lib/test/test_memoryio.py]: Hmm... I found more than that, though they're all in tests. Maybe it's just my personal preference - don't worry about it if you don't want to

Comment (orenmn) [reply Lib/test/test_memoryio.py]: not worried at all, and I would certainly follow your advice.
just wondered whether this should be added to the PEP..

and it seems that the docs think like you - https://docs.python.org/3.7/tutorial/classes.html#class-definition-syntax.

**Comment by orenmn** [comment]
I didn't expect the Spanish Inquisition!

File: Lib/_pyio.py
Comment (orenmn) [inline Lib/_pyio.py:511]: hope you are OK with this phrasing.. I chose it because even though 'int-like'
is intuitive, it seems that 'integer' is much more common in the docs and in
error messages (including in _pyio.py).

**Review by zooba — APPROVED** [review]
Yes, I'm happy with this now. Just needs the news.d item - click "Details" next to the check below and it'll give you the instructions to add this.

**Comment by zooba** [comment]
Thanks! I removed my name from the news item, since it doesn't have to go there, and once the build completes I'll merge.

**Comment by orenmn** [comment]
I added your name because most of the code in the patch was written by you.
thanks for the review :)

**Comment by zooba** [comment]
Heh, maybe, but there's a lot of code in Python that doesn't have my name on it :) That's what happens when you get merge permissions - you stop getting explicit credit for every little change.

**Pull request merged**
